### PR TITLE
feat(concerto-tools) : generate open api

### PR DESCRIPTION
**Pull request by dselman**
Signed-off-by: Dan Selman <danscode@selman.org>

Adds a code generation target to convert a Model Manager to an Open API Spec v3 document. All concept declaration with explicit identifiers and the `@resource` decorator are exposed as RESTful resources.

### Changes
- New Open API visitor
- New parameter for the existing JSON Schema visitor to allow the caller to specify how JSON schema types are stored

### Flags
- The [IBM Open API ruleset](https://www.npmjs.com/package/@ibm-cloud/openapi-ruleset) has some errors/warnings on the output but the output works with a Typescript code generator.

### Screenshots or Video

```
namespace org.acme.hr@1.0.0

enum State {
    o MA
    o NY
    o CO
    o WA
    o IL
    o CA
}

concept Address {
    o String street
    o String city
    o State state optional
    o String zipCode
    o String country
}

concept Company {
    o String name
    o Address headquarters
}

enum Department {
    o Sales
    o Marketing
    o Finance
    o HR
    o Engineering
    o Design
}

@resource
abstract asset Equipment identified by serialNumber {
    o String serialNumber
}

enum LaptopMake {
    o Apple
    o Microsoft
}

asset Laptop extends Equipment {
    o LaptopMake make
}

scalar SSN extends String default="000-00-0000" regex=/\d{3}-\d{2}-\{4}+/

@resource
abstract participant Person identified by email {
    o String email
    o String firstName
    o String lastName
    o String middleNames optional
    o Address homeAddress
    o SSN ssn
}

participant Employee extends Person {
    o String employeeId
    o Department department
    o Address officeAddress
    o Equipment[] companyAssets
    --> Manager manager optional
}

participant Contractor extends Person {
    o Company company
    --> Manager manager optional
}

participant Manager extends Employee {
    --> Person[] reports optional
}

event CompanyEvent {
}

event Onboarded extends CompanyEvent {
    --> Employee employee
}

transaction ChangeOfAddress {
    --> Person Person
    o Address newAddress
}
```

Generated OpenAPI:

```
{
  "openapi": "3.0.2",
  "servers": [],
  "info": {
    "title": "Generated Open API from Concerto Models",
    "version": "1.0.0"
  },
  "components": {
    "schemas": {
      "org.acme.hr@1.0.0.State": {
        "title": "State",
        "description": "An instance of org.acme.hr@1.0.0.State",
        "enum": [
          "MA",
          "NY",
          "CO",
          "WA",
          "IL",
          "CA"
        ]
      },
      "org.acme.hr@1.0.0.Address": {
        "title": "Address",
        "description": "An instance of org.acme.hr@1.0.0.Address",
        "type": "object",
        "properties": {
          "$class": {
            "type": "string",
            "default": "org.acme.hr@1.0.0.Address",
            "pattern": "^org\\.acme\\.hr@1\\.0\\.0\\.Address$",
            "description": "The class identifier for this type"
          },
          "street": {
            "type": "string"
          },
          "city": {
            "type": "string"
          },
          "state": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.State"
          },
          "zipCode": {
            "type": "string"
          },
          "country": {
            "type": "string"
          }
        },
        "required": [
          "$class",
          "street",
          "city",
          "zipCode",
          "country"
        ]
      },
      "org.acme.hr@1.0.0.Company": {
        "title": "Company",
        "description": "An instance of org.acme.hr@1.0.0.Company",
        "type": "object",
        "properties": {
          "$class": {
            "type": "string",
            "default": "org.acme.hr@1.0.0.Company",
            "pattern": "^org\\.acme\\.hr@1\\.0\\.0\\.Company$",
            "description": "The class identifier for this type"
          },
          "name": {
            "type": "string"
          },
          "headquarters": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.Address"
          }
        },
        "required": [
          "$class",
          "name",
          "headquarters"
        ]
      },
      "org.acme.hr@1.0.0.Department": {
        "title": "Department",
        "description": "An instance of org.acme.hr@1.0.0.Department",
        "enum": [
          "Sales",
          "Marketing",
          "Finance",
          "HR",
          "Engineering",
          "Design"
        ]
      },
      "org.acme.hr@1.0.0.Equipment": {
        "title": "Equipment",
        "description": "An instance of org.acme.hr@1.0.0.Equipment",
        "type": "object",
        "properties": {
          "$class": {
            "type": "string",
            "default": "org.acme.hr@1.0.0.Equipment",
            "pattern": "^org\\.acme\\.hr@1\\.0\\.0\\.Equipment$",
            "description": "The class identifier for this type"
          },
          "serialNumber": {
            "type": "string",
            "description": "The instance identifier for this type"
          }
        },
        "required": [
          "$class",
          "serialNumber"
        ],
        "$decorators": {
          "resource": []
        }
      },
      "org.acme.hr@1.0.0.LaptopMake": {
        "title": "LaptopMake",
        "description": "An instance of org.acme.hr@1.0.0.LaptopMake",
        "enum": [
          "Apple",
          "Microsoft"
        ]
      },
      "org.acme.hr@1.0.0.Laptop": {
        "title": "Laptop",
        "description": "An instance of org.acme.hr@1.0.0.Laptop",
        "type": "object",
        "properties": {
          "$class": {
            "type": "string",
            "default": "org.acme.hr@1.0.0.Laptop",
            "pattern": "^org\\.acme\\.hr@1\\.0\\.0\\.Laptop$",
            "description": "The class identifier for this type"
          },
          "make": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.LaptopMake"
          },
          "serialNumber": {
            "type": "string",
            "description": "The instance identifier for this type"
          }
        },
        "required": [
          "$class",
          "make",
          "serialNumber"
        ]
      },
      "org.acme.hr@1.0.0.SSN": {
        "type": "string",
        "pattern": "\\d{3}-\\d{2}-\\{4}+"
      },
      "org.acme.hr@1.0.0.Person": {
        "title": "Person",
        "description": "An instance of org.acme.hr@1.0.0.Person",
        "type": "object",
        "properties": {
          "$class": {
            "type": "string",
            "default": "org.acme.hr@1.0.0.Person",
            "pattern": "^org\\.acme\\.hr@1\\.0\\.0\\.Person$",
            "description": "The class identifier for this type"
          },
          "email": {
            "type": "string",
            "description": "The instance identifier for this type"
          },
          "firstName": {
            "type": "string"
          },
          "lastName": {
            "type": "string"
          },
          "middleNames": {
            "type": "string"
          },
          "homeAddress": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.Address"
          },
          "ssn": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.SSN"
          }
        },
        "required": [
          "$class",
          "email",
          "firstName",
          "lastName",
          "homeAddress",
          "ssn"
        ],
        "$decorators": {
          "resource": []
        }
      },
      "org.acme.hr@1.0.0.Employee": {
        "title": "Employee",
        "description": "An instance of org.acme.hr@1.0.0.Employee",
        "type": "object",
        "properties": {
          "$class": {
            "type": "string",
            "default": "org.acme.hr@1.0.0.Employee",
            "pattern": "^org\\.acme\\.hr@1\\.0\\.0\\.Employee$",
            "description": "The class identifier for this type"
          },
          "employeeId": {
            "type": "string"
          },
          "department": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.Department"
          },
          "officeAddress": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.Address"
          },
          "companyAssets": {
            "type": "array",
            "items": {
              "$ref": "#/components/schemas/org.acme.hr@1.0.0.Equipment"
            }
          },
          "manager": {
            "type": "string",
            "description": "The identifier of an instance of org.acme.hr@1.0.0.Manager"
          },
          "email": {
            "type": "string",
            "description": "The instance identifier for this type"
          },
          "firstName": {
            "type": "string"
          },
          "lastName": {
            "type": "string"
          },
          "middleNames": {
            "type": "string"
          },
          "homeAddress": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.Address"
          },
          "ssn": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.SSN"
          }
        },
        "required": [
          "$class",
          "employeeId",
          "department",
          "officeAddress",
          "companyAssets",
          "email",
          "firstName",
          "lastName",
          "homeAddress",
          "ssn"
        ]
      },
      "org.acme.hr@1.0.0.Contractor": {
        "title": "Contractor",
        "description": "An instance of org.acme.hr@1.0.0.Contractor",
        "type": "object",
        "properties": {
          "$class": {
            "type": "string",
            "default": "org.acme.hr@1.0.0.Contractor",
            "pattern": "^org\\.acme\\.hr@1\\.0\\.0\\.Contractor$",
            "description": "The class identifier for this type"
          },
          "company": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.Company"
          },
          "manager": {
            "type": "string",
            "description": "The identifier of an instance of org.acme.hr@1.0.0.Manager"
          },
          "email": {
            "type": "string",
            "description": "The instance identifier for this type"
          },
          "firstName": {
            "type": "string"
          },
          "lastName": {
            "type": "string"
          },
          "middleNames": {
            "type": "string"
          },
          "homeAddress": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.Address"
          },
          "ssn": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.SSN"
          }
        },
        "required": [
          "$class",
          "company",
          "email",
          "firstName",
          "lastName",
          "homeAddress",
          "ssn"
        ]
      },
      "org.acme.hr@1.0.0.Manager": {
        "title": "Manager",
        "description": "An instance of org.acme.hr@1.0.0.Manager",
        "type": "object",
        "properties": {
          "$class": {
            "type": "string",
            "default": "org.acme.hr@1.0.0.Manager",
            "pattern": "^org\\.acme\\.hr@1\\.0\\.0\\.Manager$",
            "description": "The class identifier for this type"
          },
          "reports": {
            "type": "array",
            "items": {
              "type": "string",
              "description": "The identifier of an instance of org.acme.hr@1.0.0.Person"
            }
          },
          "employeeId": {
            "type": "string"
          },
          "department": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.Department"
          },
          "officeAddress": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.Address"
          },
          "companyAssets": {
            "type": "array",
            "items": {
              "$ref": "#/components/schemas/org.acme.hr@1.0.0.Equipment"
            }
          },
          "manager": {
            "type": "string",
            "description": "The identifier of an instance of org.acme.hr@1.0.0.Manager"
          },
          "email": {
            "type": "string",
            "description": "The instance identifier for this type"
          },
          "firstName": {
            "type": "string"
          },
          "lastName": {
            "type": "string"
          },
          "middleNames": {
            "type": "string"
          },
          "homeAddress": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.Address"
          },
          "ssn": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.SSN"
          }
        },
        "required": [
          "$class",
          "employeeId",
          "department",
          "officeAddress",
          "companyAssets",
          "email",
          "firstName",
          "lastName",
          "homeAddress",
          "ssn"
        ]
      },
      "org.acme.hr@1.0.0.CompanyEvent": {
        "title": "CompanyEvent",
        "description": "An instance of org.acme.hr@1.0.0.CompanyEvent",
        "type": "object",
        "properties": {
          "$class": {
            "type": "string",
            "default": "org.acme.hr@1.0.0.CompanyEvent",
            "pattern": "^org\\.acme\\.hr@1\\.0\\.0\\.CompanyEvent$",
            "description": "The class identifier for this type"
          }
        },
        "required": [
          "$class"
        ]
      },
      "org.acme.hr@1.0.0.Onboarded": {
        "title": "Onboarded",
        "description": "An instance of org.acme.hr@1.0.0.Onboarded",
        "type": "object",
        "properties": {
          "$class": {
            "type": "string",
            "default": "org.acme.hr@1.0.0.Onboarded",
            "pattern": "^org\\.acme\\.hr@1\\.0\\.0\\.Onboarded$",
            "description": "The class identifier for this type"
          },
          "employee": {
            "type": "string",
            "description": "The identifier of an instance of org.acme.hr@1.0.0.Employee"
          }
        },
        "required": [
          "$class",
          "employee"
        ]
      },
      "org.acme.hr@1.0.0.ChangeOfAddress": {
        "title": "ChangeOfAddress",
        "description": "An instance of org.acme.hr@1.0.0.ChangeOfAddress",
        "type": "object",
        "properties": {
          "$class": {
            "type": "string",
            "default": "org.acme.hr@1.0.0.ChangeOfAddress",
            "pattern": "^org\\.acme\\.hr@1\\.0\\.0\\.ChangeOfAddress$",
            "description": "The class identifier for this type"
          },
          "Person": {
            "type": "string",
            "description": "The identifier of an instance of org.acme.hr@1.0.0.Person"
          },
          "newAddress": {
            "$ref": "#/components/schemas/org.acme.hr@1.0.0.Address"
          }
        },
        "required": [
          "$class",
          "Person",
          "newAddress"
        ]
      }
    }
  },
  "paths": {
    "/equipment": {
      "summary": "Path used to manage the list of equipment.",
      "description": "The REST endpoint/path used to list and create zero or more `equipment` entities.  This path contains a `GET` and `POST` operation to perform the list and create tasks, respectively.",
      "get": {
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "type": "array",
                  "items": {
                    "$ref": "#/components/schemas/org.acme.hr@1.0.0.Equipment"
                  }
                }
              }
            },
            "description": "Successful response - returns an array of `equipment` entities."
          }
        },
        "operationId": "listEquipment",
        "summary": "List All Equipment",
        "description": "Gets a list of all `equipment` entities."
      },
      "post": {
        "requestBody": {
          "description": "A new `equipment` to be created.",
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/org.acme.hr@1.0.0.Equipment"
              }
            }
          },
          "required": true
        },
        "responses": {
          "201": {
            "description": "Successful response."
          }
        },
        "operationId": "createEquipment",
        "summary": "Create a Equipment",
        "description": "Creates a new instance of a `equipment`."
      }
    },
    "/equipment/{serialNumber}": {
      "summary": "Path used to manage a single equipment.",
      "description": "The REST endpoint/path used to get, update, and delete single instances of a `equipment`.  This path contains `GET`, `PUT`, and `DELETE` operations used to perform the get, update, and delete tasks, respectively.",
      "get": {
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/org.acme.hr@1.0.0.Equipment"
                }
              }
            },
            "description": "Successful response - returns a single `equipment`."
          }
        },
        "operationId": "getEquipment",
        "summary": "Get a equipment",
        "description": "Gets the details of a single instance of a `equipment`."
      },
      "put": {
        "requestBody": {
          "description": "Updated `equipment` information.",
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/org.acme.hr@1.0.0.Equipment"
              }
            }
          },
          "required": true
        },
        "responses": {
          "202": {
            "description": "Successful response."
          }
        },
        "operationId": "replaceEquipment",
        "summary": "Update a equipment",
        "description": "Updates an existing `equipment`."
      },
      "delete": {
        "responses": {
          "204": {
            "description": "Successful response."
          }
        },
        "operationId": "deleteEquipment",
        "summary": "Delete a equipment",
        "description": "Deletes an existing `equipment`."
      },
      "parameters": [
        {
          "name": "serialNumber",
          "description": "A unique identifier for a `Equipment`.",
          "schema": {
            "type": "string"
          },
          "in": "path",
          "required": true
        }
      ]
    },
    "/people": {
      "summary": "Path used to manage the list of people.",
      "description": "The REST endpoint/path used to list and create zero or more `person` entities.  This path contains a `GET` and `POST` operation to perform the list and create tasks, respectively.",
      "get": {
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "type": "array",
                  "items": {
                    "$ref": "#/components/schemas/org.acme.hr@1.0.0.Person"
                  }
                }
              }
            },
            "description": "Successful response - returns an array of `person` entities."
          }
        },
        "operationId": "listPeople",
        "summary": "List All People",
        "description": "Gets a list of all `person` entities."
      },
      "post": {
        "requestBody": {
          "description": "A new `person` to be created.",
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/org.acme.hr@1.0.0.Person"
              }
            }
          },
          "required": true
        },
        "responses": {
          "201": {
            "description": "Successful response."
          }
        },
        "operationId": "createPerson",
        "summary": "Create a Person",
        "description": "Creates a new instance of a `person`."
      }
    },
    "/people/{email}": {
      "summary": "Path used to manage a single person.",
      "description": "The REST endpoint/path used to get, update, and delete single instances of a `person`.  This path contains `GET`, `PUT`, and `DELETE` operations used to perform the get, update, and delete tasks, respectively.",
      "get": {
        "responses": {
          "200": {
            "content": {
              "application/json": {
                "schema": {
                  "$ref": "#/components/schemas/org.acme.hr@1.0.0.Person"
                }
              }
            },
            "description": "Successful response - returns a single `person`."
          }
        },
        "operationId": "getPerson",
        "summary": "Get a person",
        "description": "Gets the details of a single instance of a `person`."
      },
      "put": {
        "requestBody": {
          "description": "Updated `person` information.",
          "content": {
            "application/json": {
              "schema": {
                "$ref": "#/components/schemas/org.acme.hr@1.0.0.Person"
              }
            }
          },
          "required": true
        },
        "responses": {
          "202": {
            "description": "Successful response."
          }
        },
        "operationId": "replacePerson",
        "summary": "Update a person",
        "description": "Updates an existing `person`."
      },
      "delete": {
        "responses": {
          "204": {
            "description": "Successful response."
          }
        },
        "operationId": "deletePerson",
        "summary": "Delete a person",
        "description": "Deletes an existing `person`."
      },
      "parameters": [
        {
          "name": "email",
          "description": "A unique identifier for a `Person`.",
          "schema": {
            "type": "string"
          },
          "in": "path",
          "required": true
        }
      ]
    }
  }
}
```

### Related Issues
- Issue #<NUMBER>
- Pull Request #<NUMBER>

### Author Checklist
- [ ] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [ ] Vital features and changes captured in unit and/or integration tests
- [ ] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [ ] Extend the documentation, if necessary
- [ ] Merging to `main` from `fork:branchname`

### CODE REVIEW

File: packages/concerto-tools/lib/codegen/fromcto/openapi/openapivisitor.js [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+        const title = parameters?.openApiTitle
+            ? parameters.openApiTitle
+            : 'Generated Open API from Concerto Models';
+        const version = parameters?.openApiVersion
+            ? parameters.openApiVersion
+            : '1.0.0';
+
+        let result = {
+            openapi: '3.0.2',
+            servers: parameters?.openApiServers ? parameters.openApiServers : [],
```
Comment (sstone1) [inline packages/concerto-tools/lib/codegen/fromcto/openapi/openapivisitor.js:83]: Could be reduced to `parameters?.openApiServers ?? []` but perhaps our level of JS doesn't allow for it.

**Review by sstone1 — APPROVED** [review]
LGTM

**Comment by mttrbrts** [comment]
Please bump the code coverage for the new visitor.

**Comment by dselman** [comment]
> Please bump the code coverage for the new visitor.

All lines covered except the handling of an unknown types by the visitor, so I think we are good.

File: packages/concerto-tools/lib/codegen/fromcto/openapi/openapivisitor.js (hunk 1 above)
Comment (dselman) [reply packages/concerto-tools/lib/codegen/fromcto/openapi/openapivisitor.js:83]: Thanks. I've created a follow up PR for that.

**Pull request merged**